spec/puppetserver/ca/sign_action_spec.rb

```diff
@@ -0,0 +1,85 @@
+  let(:err)    { StringIO.new }
+  let(:inform) { StringIO.new }
+  let(:logger) { Puppetserver::Ca::Logger.new(:info, inform, err) }
+  let(:action) { Puppetserver::Ca::SignAction.new(logger) }
+
+  describe 'validation' do
+    before do
+      allow_any_instance_of(Puppetserver::Ca::SignAction).to receive(:run).
+          and_raise("Validation succeeded when it should have failed")
+    end
+
+    it 'does not expect an argument' do
+      exit_code = Puppetserver::Ca::Cli.run(['sign','--all','foo'],STDOUT, err)
+      expect(err.string).to match(/Error:.*Unknown arguments or flags:.*foo/m)
+      expect(exit_code).to eq(1)
+    end
+    it 'does not allow --all with a valid certname call' do
+      exit_code = Puppetserver::Ca::Cli.run(['sign','--certname','foo','--all'],STDOUT,err)
+      expect(err.string).to include('--all and --certname cannot be used together')
+      expect(exit_code).to eq(1)
+    end
+    it 'errors without any arguments' do
+      exit_code = Puppetserver::Ca::Cli.run(['sign'],STDOUT,err)
+      expect(err.string).to include('No arguments given')
+      expect(exit_code).to eq(1)
+    end
+    it 'errors without a certname' do
+      exit_code = Puppetserver::Ca::Cli.run(['sign','--certname'],STDOUT,err)
+      expect(err.string).to include('A certname to sign is required')
+      expect(exit_code).to eq(1)
+    end
+    it 'does not allow --certname and --all' do
+      exit_code = Puppetserver::Ca::Cli.run(['sign','--certname','--all'],STDOUT,err)
+      expect(err.string).to match(/Cannot use --all.*HTTP API/)
+      expect(exit_code).to eq(1)
+    end
+    it 'works with a single cert' do
+      results, exit_code = action.parse(['--certname','foo'])
+      expect(results['certname']).to eq(['foo'])
+      expect(exit_code).to eq(nil)
+    end
+    it 'works with a comma separated list of certs' do
+      results, exit_code = action.parse(['--certname','foo,bar,baz'])
+      expect(results['certname']).to eq(['foo','bar','baz'])
+      expect(exit_code).to eq(nil)
+    end
+  end
+
+  describe 'error handling' do
+    let(:response)  { Struct.new(:code, :body) }
+    let(:success)   { response.new('204', nil) }
+    let(:not_found) { response.new('404', 'Not Found') }
+    let(:empty)     { response.new('404', '[]') }
+
+    it 'signs certs with successfull put' do
+      allow(action).to receive(:sign_cert).and_return(success)
+      exit_code = action.run({'certname' => ['foo']})
+      expect(exit_code).to eq(0)
+      expect(inform.string).to include('Signed certificate for foo')
+    end
+    it 'fails when put request errors' do
+      allow(action).to receive(:get_certificate_statuses).and_return(not_found)
+      exit_code = action.run({'all' => true})
+      expect(exit_code).to eq(1)
+    end
+    it 'fails when no pending certs' do
+      allow(action).to receive(:get_all_certs).and_return(empty)
+      exit_code = action.run({'all' => true})
+      expect(exit_code).to eq(1)
```

Comment:
We should run this by @Magisus when she gets back, but `puppet cert sign` returns `24` when it can't find a/any cert(s) to sign. Should maintain that weirdness or consolidate into just `1`? (We can answer that in a fixup ticket after she gets back)